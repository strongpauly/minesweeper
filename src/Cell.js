import React, { Component } from 'react';
import mine from './mine.svg';

class Cell extends Component {

  constructor(props) {
      super(props);
      this.checkCell = this.checkCell.bind(this);
      this.markCell = this.markCell.bind(this);
  }

  checkCell(e) {
      //Can't check more than once.
      if (!this.props.checked) {
          this.props.onCheck(this.props.x, this.props.y)
      }
  }

  markCell(e) {
      e.preventDefault();
      if(!this.props.checked) {
          this.props.onMark(this.props.x, this.props.y, !this.props.marked);
      }
  }

  render() {
      let className = ["cell"];
      let content = " ";
      if (this.props.checked) {
          className.push("checked");
          if (this.props.adjacentCount !== undefined) {
              content = this.props.adjacentCount;
              className.push("count-" + this.props.adjacentCount);
          }
      } else if (this.props.marked) {
          className.push("marked");
          content = "F";
      }
      if(this.props.mine) {
          className.push("mine");
          if(this.props.marked) {
              content = "X"
          } else {
              content = <img src={mine} className="mineImg" alt="mine" />;
          }
      }
      return <td className={ className.join(" ") } onClick={this.checkCell} onContextMenu={this.markCell}>{ content }</td>
  }
}

export default Cell;
