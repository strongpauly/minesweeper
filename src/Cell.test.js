import React from 'react';
import Cell from './Cell';
import {mount, shallow} from 'enzyme';
import sinon from 'sinon';

/* eslint-env jest */

describe('<Cell>', () => {

  it('renders without crashing', () => {
    shallow(<Cell x={0} y={0}/>);
  });

  it('has "cell" css class', () => {
    const cell = shallow(<Cell x={0} y={0}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('.cell')).toHaveLength(1);
  });

  it('initialised with checked has "checked" class', () => {
    const cell = shallow(<Cell x={0} y={0} checked={true}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('.checked')).toHaveLength(1);
  });

  it('initialised with mine has "mine" class', () => {
    const cell = shallow(<Cell x={0} y={0} mine={true}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('.mine')).toHaveLength(1);
  });

  it('initialised with mine shows mine img', () => {
    const cell = shallow(<Cell x={0} y={0} mine={true}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('img')).toHaveLength(1);
  });

  it('initialised with marked has "marked" class', () => {
    const cell = shallow(<Cell x={0} y={0} marked={true}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('.marked')).toHaveLength(1);
  });

  it('with a marked mine show an X', () => {
    const cell = shallow(<Cell x={0} y={0} mine={true} marked={true}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('td').text()).toEqual('X');
  });

  it('initialised with an adjacent count, renders that count with "count-" class', () => {
    const cell = shallow(<Cell x={0} y={0} checked={true} adjacentCount={1}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('.count-1')).toHaveLength(1);
    expect(cell.find('td').text()).toEqual('1');
  });

  it('click fires onCheck handler', () => {
    const onCheck = sinon.spy();
    let table = mount(<table><tbody><tr><Cell x={0} y={0} onCheck={onCheck}/></tr></tbody></table>);
    table.find('td').simulate('click');
    expect(onCheck.calledOnce).toEqual(true);
  });

  it('context menu fires onMarked handler', () => {
    const onMark = sinon.spy();
    let table = mount(<table><tbody><tr><Cell x={0} y={0} onMark={onMark}/></tr></tbody></table>);
    table.find('td').simulate('contextMenu');
    expect(onMark.calledOnce).toEqual(true);
  });

  it('click won\'t fire onCheck if checked is true', () => {
    const onCheck = sinon.spy();
    let table = mount(<table><tbody><tr><Cell x={0} y={0} onCheck={onCheck} checked={true}/></tr></tbody></table>);
    let td = table.find('td');
    td.simulate('click');
    expect(onCheck.calledOnce).toEqual(false);
  });

  it('click won\'t fire onCheck if marked is true', () => {
    const onCheck = sinon.spy();
    let table = mount(<table><tbody><tr><Cell x={0} y={0} onCheck={onCheck} marked={true}/></tr></tbody></table>);
    let td = table.find('td');
    td.simulate('click');
    expect(onCheck.calledOnce).toEqual(false);
  });

  it('click won\'t fire onMark if checked is true', () => {
     const onMark = sinon.spy();
    let table = mount(<table><tbody><tr><Cell x={0} y={0} onMark={onMark} checked={true}/></tr></tbody></table>);
    table.find('td').simulate('contextMenu');
    expect(onMark.calledOnce).toEqual(false);
  });
});
