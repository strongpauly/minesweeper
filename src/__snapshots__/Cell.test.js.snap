// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell has "cell" css class 1`] = `
<td
  className="cell"
  onClick={[Function]}
  onContextMenu={[Function]}
>
   
</td>
`;

exports[`Cell initialised with checked has "checked" class 1`] = `
<td
  className="cell checked"
  onClick={[Function]}
  onContextMenu={[Function]}
>
   
</td>
`;

exports[`Cell initialised with marked has "marked" class 1`] = `
<td
  className="cell marked"
  onClick={[Function]}
  onContextMenu={[Function]}
>
  !
</td>
`;

exports[`Cell initialised with mine has "mine" class 1`] = `
<td
  className="cell mine"
  onClick={[Function]}
  onContextMenu={[Function]}
>
  <img
    alt="mine"
    className="mineImg spinning"
    src="mine.svg"
  />
</td>
`;

exports[`Cell initialised with mine shows mine img 1`] = `
<td
  className="cell mine"
  onClick={[Function]}
  onContextMenu={[Function]}
>
  <img
    alt="mine"
    className="mineImg spinning"
    src="mine.svg"
  />
</td>
`;
